File: packages/python/src/memcell/scoped.py
import contextlib
from collections.abc import Awaitable, Callable
from typing import Any, Generic, TypeVar

from pydantic import BaseModel

from .models import (
    FeedbackResponse,
    JobEvent,
    OutcomeVerdict,
    RecallResponse,
    RememberResponse,
    ReportResponse,
    ScopedExecutionContext,
)

T = TypeVar("T")


class ScopedExecutionResult(BaseModel, Generic[T]):
    """Result envelope for wrapped agent execution."""

    result: T
    report: ReportResponse
    recall: RecallResponse


class ScopedMemCell:
    """Synchronous memory handle scoped to a specific project namespace and default subject."""

    def __init__(
        self,
        client: Any,
        namespace: str,
        subject: str | None = None,
    ) -> None:
        self._client = client
        self.namespace = namespace
        self.default_subject = subject

    def recall(
        self,
        query: str,
        subject: str | None = None,
        kind: str | list[str] | None = None,
        min_confidence: float | None = None,
        limit: int | None = None,
        tags: list[str] | None = None,
        format: str = "xml",
        allow_provisional: bool | None = None,
    ) -> RecallResponse:
        return self._client.recall(
            query=query,
            namespace=self.namespace,
            subject=subject or self.default_subject,
            kind=kind,
            min_confidence=min_confidence,
            limit=limit,
            tags=tags,
            format=format,
            allow_provisional=allow_provisional,
        )

    def remember(
        self,
        title: str | None = None,
        context: str | None = None,
        example: str | None = None,
        tags: list[str] | None = None,
        subject: str | None = None,
        kind: str | None = None,
        status: str | None = None,
        confidence: float | None = None,
        expires_at: Any | None = None,
        raw: str | None = None,
        session_id: str | None = None,
        async_: bool | None = None,
    ) -> RememberResponse:
        return self._client.remember(
            title=title,
            context=context,
            example=example,
            tags=tags,
            subject=subject or self.default_subject,
            kind=kind,
            status=status,
            confidence=confidence,
            expires_at=expires_at,
            raw=raw,
            session_id=session_id,
            namespace=self.namespace,
            async_=async_,
        )

    def report(
        self,
        action_taken: str,
        outcome: OutcomeVerdict,
        subject: str | None = None,
        reason: str | None = None,
        external_ref: str | None = None,
        payload: dict[str, Any] | None = None,
        recall_id: str | None = None,
        statement_id: str | None = None,
        auto_distill: bool = True,
        async_: bool | None = None,
    ) -> ReportResponse:
        return self._client.report(
            action_taken=action_taken,
            outcome=outcome,
            namespace=self.namespace,
            subject=subject or self.default_subject,
            reason=reason,
            external_ref=external_ref,
            payload=payload,
            recall_id=recall_id,
            statement_id=statement_id,
            auto_distill=auto_distill,
            async_=async_,
        )

    def feedback(
        self,
        statement_id: str,
        outcome: OutcomeVerdict,
        recall_id: str | None = None,
        reason: str | None = None,
        external_ref: str | None = None,
        payload: dict[str, Any] | None = None,
    ) -> FeedbackResponse:
        return self._client.feedback(
            statement_id=statement_id,
            outcome=outcome,
            recall_id=recall_id,
            reason=reason,
            external_ref=external_ref,
            payload=payload,
            namespace=self.namespace,
        )

    def wait_for_job(
        self,
        job_id: str,
        timeout_ms: int = 15000,
        poll_interval_ms: int = 250,
        on_progress: Callable[[JobEvent], None] | None = None,
    ) -> JobEvent:
        return self._client.wait_for_job(
            job_id=job_id,
            timeout_ms=timeout_ms,
            poll_interval_ms=poll_interval_ms,
            on_progress=on_progress,
        )

    def wrap_execution(
        self,
        action: str,
        fn: Callable[[ScopedExecutionContext], T],
        subject: str | None = None,
        external_ref: str | None = None,
        payload: dict[str, Any] | None = None,
        query: str | None = None,
    ) -> ScopedExecutionResult[T]:
        target_subject = subject or self.default_subject
        recall = self.recall(query=query or action, subject=target_subject)
        context = ScopedExecutionContext(
            recall_id=recall.recall_id,
            prompt_context=recall.prompt_context,
            statements=recall.statements,
            action=action,
            subject=target_subject,
        )

        try:
            result = fn(context)
            report = self.report(
                action_taken=action,
                outcome="worked",
                subject=target_subject,
                recall_id=recall.recall_id,
                external_ref=external_ref,
                payload=payload,
            )
            return ScopedExecutionResult[T](result=result, report=report, recall=recall)
        except Exception as e:
            with contextlib.suppress(Exception):
                self.report(
                    action_taken=action,
                    outcome="failed",
                    reason=str(e),
                    subject=target_subject,
                    recall_id=recall.recall_id,
                    external_ref=external_ref,
                    payload=payload,
                )
            raise


class AsyncScopedMemCell:
    """Asynchronous memory handle scoped to a specific project namespace and default subject."""

    def __init__(
        self,
        client: Any,
        namespace: str,
        subject: str | None = None,
    ) -> None:
        self._client = client
        self.namespace = namespace
        self.default_subject = subject

    async def recall(
        self,
        query: str,
        subject: str | None = None,
        kind: str | list[str] | None = None,
        min_confidence: float | None = None,
        limit: int | None = None,
        tags: list[str] | None = None,
        format: str = "xml",
        allow_provisional: bool | None = None,
    ) -> RecallResponse:
        return await self._client.recall(
            query=query,
            namespace=self.namespace,
            subject=subject or self.default_subject,
            kind=kind,
            min_confidence=min_confidence,
            limit=limit,
            tags=tags,
            format=format,
            allow_provisional=allow_provisional,
        )

    async def remember(
        self,
        title: str | None = None,
        context: str | None = None,
        example: str | None = None,
        tags: list[str] | None = None,
        subject: str | None = None,
        kind: str | None = None,
        status: str | None = None,
        confidence: float | None = None,
        expires_at: Any | None = None,
        raw: str | None = None,
        session_id: str | None = None,
        async_: bool | None = None,
    ) -> RememberResponse:
        return await self._client.remember(
            title=title,
            context=context,
            example=example,
            tags=tags,
            subject=subject or self.default_subject,
            kind=kind,
            status=status,
            confidence=confidence,
            expires_at=expires_at,
            raw=raw,
            session_id=session_id,
            namespace=self.namespace,
            async_=async_,
        )

    async def report(
        self,
        action_taken: str,
        outcome: OutcomeVerdict,
        subject: str | None = None,
        reason: str | None = None,
        external_ref: str | None = None,
        payload: dict[str, Any] | None = None,
        recall_id: str | None = None,
        statement_id: str | None = None,
        auto_distill: bool = True,
        async_: bool | None = None,
    ) -> ReportResponse:
        return await self._client.report(
            action_taken=action_taken,
            outcome=outcome,
            namespace=self.namespace,
            subject=subject or self.default_subject,
            reason=reason,
            external_ref=external_ref,
            payload=payload,
            recall_id=recall_id,
            statement_id=statement_id,
            auto_distill=auto_distill,
            async_=async_,
        )

    async def feedback(
        self,
        statement_id: str,
        outcome: OutcomeVerdict,
        recall_id: str | None = None,
        reason: str | None = None,
        external_ref: str | None = None,
        payload: dict[str, Any] | None = None,
    ) -> FeedbackResponse:
        return await self._client.feedback(
            statement_id=statement_id,
            outcome=outcome,
            recall_id=recall_id,
            reason=reason,
            external_ref=external_ref,
            payload=payload,
            namespace=self.namespace,
        )

    async def wait_for_job(
        self,
        job_id: str,
        timeout_ms: int = 15000,
        poll_interval_ms: int = 250,
        on_progress: Callable[[JobEvent], None] | None = None,
    ) -> JobEvent:
        return await self._client.wait_for_job(
            job_id=job_id,
            timeout_ms=timeout_ms,
            poll_interval_ms=poll_interval_ms,
            on_progress=on_progress,
        )

    async def wrap_execution(
        self,
        action: str,
        fn: Callable[[ScopedExecutionContext], T | Awaitable[T]],
        subject: str | None = None,
        external_ref: str | None = None,
        payload: dict[str, Any] | None = None,
        query: str | None = None,
    ) -> ScopedExecutionResult[T]:
        target_subject = subject or self.default_subject
        recall = await self.recall(query=query or action, subject=target_subject)
        context = ScopedExecutionContext(
            recall_id=recall.recall_id,
            prompt_context=recall.prompt_context,
            statements=recall.statements,
            action=action,
            subject=target_subject,
        )

        try:
            res = fn(context)
            if hasattr(res, "__await__"):
                result = await res  # type: ignore
            else:
                result = res  # type: ignore

            report = await self.report(
                action_taken=action,
                outcome="worked",
                subject=target_subject,
                recall_id=recall.recall_id,
                external_ref=external_ref,
                payload=payload,
            )
            return ScopedExecutionResult[T](result=result, report=report, recall=recall)
        except Exception as e:
            with contextlib.suppress(Exception):
                await self.report(
                    action_taken=action,
                    outcome="failed",
                    reason=str(e),
                    subject=target_subject,
                    recall_id=recall.recall_id,
                    external_ref=external_ref,
                    payload=payload,
                )
            raise

File: packages/typescript/src/scoped.ts
import type {
  FeedbackParams,
  FeedbackResponse,
  JobEvent,
  RecallParams,
  RecallResponse,
  RememberParams,
  RememberResponse,
  ReportParams,
  ReportResponse,
  ScopeOptions,
  ScopedExecutionContext,
  ScopedExecutionResult,
  WaitForJobOptions,
  WrapExecutionOptions,
} from "./types.js";
import type { MemCell } from "./client.js";

export class ScopedMemCell {
  readonly defaultSubject: string | null;
  readonly defaultFormat: "xml" | "markdown" | "none";

  constructor(
    private readonly client: MemCell,
    readonly namespace: string,
    options?: ScopeOptions,
  ) {
    this.defaultSubject = options?.subject ?? null;
    this.defaultFormat = options?.format ?? "xml";
  }

  /**
   * Pre-flight recall from the scoped memory container.
   */
  async recall(
    query: string,
    options?: Omit<RecallParams, "namespace" | "query">,
  ): Promise<RecallResponse> {
    return await this.client.recall({
      namespace: this.namespace,
      query,
      subject:
        options?.subject !== undefined ? options.subject : this.defaultSubject,
      format: options?.format ?? this.defaultFormat,
      ...options,
    });
  }

  /**
   * Direct addition of durable statements to the scoped memory container.
   */
  async remember(
    statement: Omit<RememberParams, "namespace">,
  ): Promise<RememberResponse> {
    return await this.client.remember({
      namespace: this.namespace,
      subject:
        statement.subject !== undefined
          ? statement.subject
          : this.defaultSubject,
      ...statement,
    });
  }

  /**
   * Post-flight execution reporting & dynamic reflex reinforcement.
   */
  async report(
    params: Omit<ReportParams, "namespace">,
  ): Promise<ReportResponse> {
    return await this.client.report({
      namespace: this.namespace,
      subject:
        params.subject !== undefined ? params.subject : this.defaultSubject,
      ...params,
    });
  }

  /**
   * Direct statement outcome feedback.
   */
  async feedback(
    params: Omit<FeedbackParams, "namespace">,
  ): Promise<FeedbackResponse> {
    return await this.client.feedback({
      namespace: this.namespace,
      ...params,
    });
  }

  /**
   * Automated execution wrapper implementing the complete Dynamic Agentic Loop:
   * 1. Pre-Flight Recall: Queries relevant invariants and learned reflexes for the action.
   * 2. In-Flight Execution: Runs the agent callback with promptContext and statements.
   * 3. Post-Flight Reinforcement: Automatically reports outcome ('worked' on resolution, 'failed' on exception)
   *    and attributes feedback before re-throwing any error.
   */
  async wrapExecution<T>(
    options: WrapExecutionOptions,
    actionFn: (context: ScopedExecutionContext) => Promise<T>,
  ): Promise<ScopedExecutionResult<T>> {
    const targetSubject =
      options.subject !== undefined ? options.subject : this.defaultSubject;
    const format = options.format ?? this.defaultFormat;

    // 1. Pre-Flight Recall
    const recallResult = await this.recall(options.action, {
      subject: targetSubject,
      kind: options.kind,
      minConfidence: options.minConfidence,
      limit: options.limit,
      tags: options.tags,
      format,
    });

    const executionContext: ScopedExecutionContext = {
      promptContext: recallResult.promptContext,
      statements: recallResult.statements,
      recallId: recallResult.recallId,
      subject: targetSubject,
    };

    // 2. In-Flight Execution
    let result: T;
    try {
      result = await actionFn(executionContext);
    } catch (err) {
      // 3a. Post-Flight Report on Error
      const errorMessage = err instanceof Error ? err.message : String(err);
      try {
        await this.report({
          actionTaken: options.action,
          outcome: "failed",
          recallId: recallResult.recallId,
          subject: targetSubject,
          reason: errorMessage,
          externalRef: options.externalRef,
          payload: options.payload,
          autoDistill: options.autoDistill,
        });
      } catch {
        // Do not suppress the original execution exception
      }
      throw err;
    }

    // 3b. Post-Flight Report on Success
    const reportResult = await this.report({
      actionTaken: options.action,
      outcome: "worked",
      recallId: recallResult.recallId,
      subject: targetSubject,
      externalRef: options.externalRef,
      payload: options.payload,
      autoDistill: options.autoDistill,
    });

    return {
      result,
      report: reportResult,
      recallId: recallResult.recallId,
      statements: recallResult.statements,
      promptContext: recallResult.promptContext,
    };
  }

  /**
   * Waits for an asynchronous background job to complete.
   */
  async waitForJob(
    jobId: string,
    options?: WaitForJobOptions,
  ): Promise<JobEvent> {
    return await this.client.waitForJob(jobId, options);
  }
}
